File: app/src/main/java/com/video/test/module/download/DownloadModel.java
package com.video.test.module.download;

import com.video.test.TestApp;
import com.video.test.sp.SpUtils;
import com.video.test.utils.RxSchedulers;
import com.video.test.utils.SDCardUtils;

import io.reactivex.Observable;
import io.reactivex.annotations.NonNull;
import io.reactivex.functions.BiFunction;
import io.reactivex.functions.Function;

/**
 * @author Enoch Created on 2018/6/27.
 */
public class DownloadModel implements DownloadContract.Model {
    @Override
    public Observable<Long> getSDCardFreeSize() {
        String cachePath = SpUtils.getString(TestApp.getContext(), "cachePath", SDCardUtils.getSDRootPath());
        return Observable.just(SDCardUtils.getFreeSpaceBytes(cachePath))
                .compose(RxSchedulers.io_main());
    }

    @Override
    public Observable<String> getSDCardTotalSize() {
        String cachePath = SpUtils.getString(TestApp.getContext(), "cachePath", SDCardUtils.getSDRootPath());
        return Observable.just(SDCardUtils.getTotalSpaceBytes(cachePath))
                .map(new Function<Long, String>() {
                    @Override
                    public String apply(@NonNull Long aLong) {
                        return SDCardUtils.getSizeString(aLong);
                    }
                }).compose(RxSchedulers.io_main());
    }

    @Override
    public Observable<Integer> getSDCardUsedPercent() {
        String cachePath = SpUtils.getString(TestApp.getContext(), "cachePath", SDCardUtils.getSDRootPath());
        Observable<Long> freeSpace = Observable.just(SDCardUtils.getFreeSpaceBytes(cachePath));
        Observable<Long> totalSpace = Observable.just(SDCardUtils.getTotalSpaceBytes(cachePath));

        return Observable.zip(totalSpace, freeSpace, new BiFunction<Long, Long, Integer>() {
            @Override
            public Integer apply(@NonNull Long aLong, @NonNull Long aLong2) {
                return (int) ((double) (aLong - aLong2) / aLong * 100.0);
            }
        });
    }


}

File: app/src/main/java/com/video/test/module/download/DownloadModel.kt
package com.video.test.module.download

import android.content.Context
import com.video.test.db.DBManager
import com.video.test.javabean.M3U8DownloadBean
import com.video.test.javabean.SdCardInfoBean
import com.video.test.utils.RxSchedulers
import com.video.test.utils.SDCardUtils
import io.reactivex.Observable
import io.reactivex.ObservableOnSubscribe

class DownloadModel : DownloadContract.Model {

    override fun getSDCardInfo(): Observable<SdCardInfoBean> {
        return Observable.create(ObservableOnSubscribe<SdCardInfoBean> { emitter ->
            val freeSpaceBytes = SDCardUtils.getFreeSpaceBytes(SDCardUtils.getSDRootPath())
            val totalSpaceBytes = SDCardUtils.getTotalSpaceBytes(SDCardUtils.getSDRootPath())
            emitter.onNext(SdCardInfoBean(freeSpaceBytes, totalSpaceBytes))
            emitter.onComplete()
        }).compose(RxSchedulers.io_main())
    }

    override fun getAllTasks(context: Context): Observable<MutableList<M3U8DownloadBean>> {
        return Observable.create { emitter ->
            val tasks = DBManager.getInstance(context).queryM3U8Tasks()
            emitter.onNext(tasks)
            emitter.onComplete()
        }
    }
}